contrib/zenith/pagestore_prefetch.c

```diff
@@ -0,0 +1,401 @@
+				entry->wait_latch = MyLatch;
+
+				LWLockRelease(prefetch_lock);
+
+				/* wait latch to be signaled */
+				(void)WaitLatch(MyLatch, WL_EXIT_ON_PM_DEATH|WL_LATCH_SET, 0, PG_WAIT_EXTENSION);
+				ResetLatch(MyLatch);
+			}
+			continue;
+		}
+		LWLockRelease(prefetch_lock);
+		return false;
+	}
+}
+
+static void
+start_prefetch_worker()
+{
+	/* Lazy start of prefetch bacnhround worker */
+	BackgroundWorker worker;
+	BackgroundWorkerHandle *handle;
+	pid_t bgw_pid;
+
+	MemSet(&worker, 0, sizeof(worker));
+	strncpy(worker.bgw_name, "zenith_prefetch", sizeof(worker.bgw_name));
+	strncpy(worker.bgw_type, "zenith_prefetch", sizeof(worker.bgw_type));
+	worker.bgw_flags = BGWORKER_SHMEM_ACCESS;
+	worker.bgw_start_time = BgWorkerStart_ConsistentState;
+	worker.bgw_restart_time = BGW_DEFAULT_RESTART_INTERVAL;
+	strcpy(worker.bgw_function_name, "zenith_prefetch_main");
+	strcpy(worker.bgw_library_name, "zenith");
+	worker.bgw_notify_pid = MyProcPid;
+	if (!RegisterDynamicBackgroundWorker(&worker, &handle))
+	{
+		elog(ERROR, "zenith: failed to start prefetch background worker");
+	}
+	if (WaitForBackgroundWorkerStartup(handle, &bgw_pid) != BGWH_STARTED)
+	{
+		elog(ERROR, "zenith: startup of prefetch background worker is failed");
+	}
+	for (int n_attempts = 0; prefetch_control->go_latch == NULL && n_attempts < 100; n_attempts++)
+	{
+		pg_usleep(10000); /* wait background worker to be registered in procarray */
+	}
+	if (prefetch_control->go_latch == NULL)
+	{
+		elog(ERROR, "zenith: background prefetch worker %d is crashed", bgw_pid);
+	}
+}
+
+void
+zenith_prefetch_buffer(SMgrRelation reln, ForkNumber forknum, BlockNumber blocknum, XLogRecPtr lsn)
+{
+	PrefetchEntry* entry;
+	bool found;
+	BufferTag tag;
+
+	if (prefetch_buffer_size == 0
+		|| reln->smgr_rnode.node.dbNode == 0
+		|| reln->smgr_rnode.node.relNode < FirstNormalObjectId)
+	{
+		/* disable prefetch for system tables */
+		return;
+	}
+
+	tag.rnode = reln->smgr_rnode.node;
+	tag.forkNum = forknum;
+	tag.blockNum = blocknum;
+
+	LWLockAcquire(prefetch_lock, LW_EXCLUSIVE);
+	entry = hash_search(prefetch_hash, &tag, HASH_ENTER, &found);
+
+	if (!found)
+	{
+		size_t curr = prefetch_control->curr;
+		PrefetchEntry* victim = prefetch_control->entries[curr];
+		if (victim != NULL)
+		{
+			/* remove old entry */
+			if (victim->wait_latch)
+				SetLatch(victim->wait_latch);
+			hash_search(prefetch_hash, &victim->tag, HASH_REMOVE, NULL);
+		}
+		entry->state = QUEUED;
+		entry->wait_latch = NULL;
+		entry->lsn = lsn;
+		entry->index = curr;
+		prefetch_control->entries[curr] = entry;
+		prefetch_control->curr = (curr + 1) % prefetch_buffer_size;
+		n_prefetch_requests += 1;
+		prefetch_log("%lu: prefetch request for block %d of relation %d",
+					 GetCurrentTimestamp(), blocknum, tag.rnode.relNode);
+
+		if (prefetch_control->go_latch == NULL)
+		{
+			/* lazy start of prefetch background worker */
+			start_prefetch_worker();
+		}
```

Comment:
I prefer to loose here some level of encapsulation but set latch out of critical section (prefetch_lock). It may reduce extra context switches when SetLatch gives control to the waken backend, but then it has to sleep because lock is still hold.